src/main/java/net/openhft/chronicle/queue/impl/single/TSQueueLock.java

```diff
@@ -0,0 +1,128 @@
+ *
+ * http://www.higherfrequencytrading.com
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package net.openhft.chronicle.queue.impl.single;
+
+import net.openhft.chronicle.core.values.LongValue;
+import net.openhft.chronicle.queue.impl.TableStore;
+import net.openhft.chronicle.queue.impl.table.SingleTableBuilder;
+import net.openhft.chronicle.threads.Pauser;
+import net.openhft.lang.Jvm;
+
+import java.io.File;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.function.Supplier;
+
+/**
+ * Implements queue lock via TableStore mechanism.
+ */
+public class TSQueueLock implements QueueLock {
+
+    private static final String QUEUE_LOCK_FILE = "queue-lock" + SingleTableBuilder.SUFFIX;
+    private static final String LOCK_KEY = "chronicle.queue.lock";
+    private static final long LOCK_WAIT_TIMEOUT = Long.getLong("chronicle.queue.lock.timeoutMS", 30_000);
+    private static final long UNLOCKED = Long.MIN_VALUE;
+    private static final long PID = Jvm.getProcessId();
+    private final LongValue lock;
+    private final Pauser pauser;
+    private final ThreadLocal<Long> lockHolderTidTL = new ThreadLocal<>();
+
+    public TSQueueLock(File queueDirectoryPath, Supplier<Pauser> pauser) {
+        final File storeFilePath;
+        if ("".equals(queueDirectoryPath.getPath())) {
+            storeFilePath = new File(QUEUE_LOCK_FILE);
+        } else {
+            storeFilePath = new File(queueDirectoryPath, QUEUE_LOCK_FILE);
+            queueDirectoryPath.mkdirs();
+        }
+        TableStore tableStore = SingleTableBuilder.binary(storeFilePath).build();
+        this.lock = tableStore.acquireValueFor(LOCK_KEY);
+        this.pauser = pauser.get();
+    }
+
+    /**
+     * Stores current PID to table store, and any other process trying to acquire lock will wait for
+     * <code>chronicle.queue.lock.timeoutMS</code> millis (default is 30000) for the lock to be released, and if it is not
+     * after timeout, throws {@link IllegalStateException}. Also the locking thread ID is stored in threadLocal field, so
+     * that only locking thread is allowed to {@link net.openhft.chronicle.queue.ChronicleQueue#acquireAppender} while the queue is locked.
+     */
+    @Override
+    public void acquireLock() {
+        long tid = Thread.currentThread().getId();
+        try {
+            while (!lock.compareAndSwapValue(UNLOCKED, PID)) {
```

Comment:
Sorry, I must be being daft - how does this exit if the current process fails to lock the queue? I can't fathom where the TimeoutException is thrown from.

Reply:
pauser.pause throws TimeoutException

Reply:
Oh, the Pauser is stateful - that's delightful.

Probably not safe for access from multiple threads then...

Reply:
Just like 99% of queue code :) Not sure we really want to handle multithreading here - essentially there shouldn't be anyone trying to acquire lock other then our own replication Source.